### PR TITLE
[replication] EOS 25355: Improvements in putBucketReplication API

### PR DESCRIPTION
**Pull request by trshaffer**
# Problem Statement
- Problem statement

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

**Comment by cortx-admin** [comment]
Can one of the admins verify this patch?

File: server/s3_put_bucket_replication_action.h [hunk 1]
```diff
@@ -90,6 +90,7 @@ class S3PutBucketReplicationAction : public S3BucketAction {
   FRIEND_TEST(S3PutBucketReplicationActionTest, SendResponseToClientSuccess);
   FRIEND_TEST(S3PutBucketReplicationActionTest,
               SendResponseToClientInternalError);
+  FRIEND_TEST(S3PutBucketReplicationActionTest, ValidateDestinationIsNotSource);
```
Comment (shraddhaghatol) [inline server/s3_put_bucket_replication_action.h:93]: Change the test name.

File: ut/s3_put_bucket_replication_action_test.cc [hunk 2]
```diff
@@ -456,3 +464,49 @@ TEST_F(S3PutBucketReplicationActionTest, SendResponseToClientInternalError) {
   EXPECT_CALL(*request_mock, send_response(500, _)).Times(AtLeast(1));
   action_under_test_ptr->send_response_to_s3_client();
 }
+
+// The source and destination buckets cannot be the same
+TEST_F(S3PutBucketReplicationActionTest, ValidateDestinationIsNotSource) {
+
+  MockReplicationConfigStr =
```
Comment (shraddhaghatol) [inline ut/s3_put_bucket_replication_action_test.cc:471]: can we write this string in following way -
MockReplicationConfigStr =
 "	<ReplicationConfiguration 													"
 " 	xmlns=\"http://s3.amazonaws.com/doc/2006-03-01/\">							"
 "	<Role>role-string</Role> 													"
 "  <Rule>																	"
 "  <Status>Enabled</Status>													"
 "	<Priority>1</Priority> 														"
 "   <ID>Rule-1</ID>															"
 "	<DeleteMarkerReplication><Status>Disabled</Status></DeleteMarkerReplication>        "
 "	<Destination><Bucket>seagatebucket</Bucket></Destination>			               "
 "	<Filter><And><Prefix/>			                                                                               "
 "	<Tag><Key>key12</Key><Value>vv</Value></Tag>								"
 "	<Tag><Key>key2</Key><Value>value2</Value></Tag>							"
 "	</And></Filter></Rule>													"
 "	<Rule>																	"
 "	<Status>Enabled</Status>													"
 " 	<Priority>2</Priority>														"
 "	<ID>Rule-2</ID>															"
 "	<DeleteMarkerReplication><Status>Disabled</Status></DeleteMarkerReplication>         "										
 "	<Destination><Bucket>seagatebucket</Bucket></Destination>					"
 "	<Filter><And><Prefix/>													"
 "	<Tag><Key>key12</Key><Value>vv</Value></Tag>								"
 "	<Tag><Key>key2</Key><Value>value2</Value></Tag>							"
 "	</And></Filter></Rule>													"
 "	</ReplicationConfiguration>													";

File: ut/s3_put_bucket_replication_action_test.cc [hunk 3]
```diff
@@ -456,3 +464,49 @@ TEST_F(S3PutBucketReplicationActionTest, SendResponseToClientInternalError) {
...
+      "<ReplicationConfiguration "
+      "xmlns=\"http://s3.amazonaws.com/doc/2006-03-01/\">"
+      "<Role>role-string</Role><Rule><Status>Enabled</Status>"
+      "<Priority>1</Priority><ID>Rule-1</ID><DeleteMarkerReplication>"
+      "<Status>Disabled</Status></DeleteMarkerReplication><Destination>"
+      "<Bucket>seagatebucket</Bucket></Destination><Filter><And><Prefix/>"
```
Comment (shraddhaghatol) [inline ut/s3_put_bucket_replication_action_test.cc:477]: <Prefix> tag is incomplete here. Please check the string once.

File: server/s3_put_bucket_replication_action.h (hunk 1 above)
Comment (trshaffer) [reply server/s3_put_bucket_replication_action.h]: Fixed

File: ut/s3_put_bucket_replication_action_test.cc (hunk 3 above)
Comment (trshaffer) [reply ut/s3_put_bucket_replication_action_test.cc:478]: Can you clarify here? I just copied from the successful test above, changing the bucket names to `seagatebucket`

File: server/s3_put_bucket_replication_action.h (hunk 1 above)
Comment (shraddhaghatol) [reply server/s3_put_bucket_replication_action.h]: Ok.

File: ut/s3_put_bucket_replication_action_test.cc (hunk 3 above)
Comment (shraddhaghatol) [reply ut/s3_put_bucket_replication_action_test.cc:478]: </Prefix> xml tag looks like incomplete here. There is only closing prefix tag ,there is no opening tag present for Prefix.

Comment (trshaffer) [reply ut/s3_put_bucket_replication_action_test.cc]: Oh yeah, it's just the self-closing form. I just switched them to open/close tags instead

**Review by mbcortx — APPROVED** [review]

File: ut/s3_put_bucket_replication_action_test.cc (hunk 3 above)
Comment (shraddhaghatol) [reply ut/s3_put_bucket_replication_action_test.cc]: okay.

**Review by shraddhaghatol — APPROVED** [review]
Looks good!!

**Pull request merged**